### PR TITLE
Enforce correct semantics for TabNav

### PR DESCRIPTION
**Pull request by owenniblock**
Enforces the containing `div` to have `role="tablist"` and the item elements to have `role="tab"` - tabs should avoid having focusable elements as children.

### Screenshots

Before:
Before screenshot showing 3 tabs

After:
After screenshot showing 3 tabs that are styled the same

### Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation
- [ ] Tested in Chrome
- [ ] Tested in Firefox
- [x] Tested in Safari
- [x] Tested in Edge

### CODE REVIEW

**Comment by siddharthkp** [comment]
Hi! Just to let you know that CI/deployments breaking is not because of the changes in this PR and is something on our radar! https://github.com/primer/react/pull/2123

**Comment by owenniblock** [comment]
Was chatting with Joel last night and now have a better handle on what `As` does. Maybe we *don't* want to support `As` here and we should have the top level HTML starting with the `<nav>`.

I also wonder whether we really want to support `sx` functionality (I see those tests disabled on a lot of other components).

As these are breaking changes, I wanted to get some feedback before pushing up a change in that direction.

**Comment by siddharthkp** [comment]
Hi! I haven't had the chance to look at this PR yet, I will come back to it with opinions/suggestions this week :)

tl;dr: 
- we want to support sx on every component
- we try to avoid breaking changes, but if there's a good case for them, we can schedule it for the next-major

**Comment by owenniblock** [comment]
Thanks @siddharthkp - my opinions below 🙂 

> it might be okay to skip the as prop here if we want to enforce the container is a `nav`

I think we do, it makes sense for TabNav to enforce the `nav` element to allow screen readers to navigate to the tabs in navigation mode. So if I try to support `As` - all I'm doing is wrapping it in a `div` that's not going to be required 99% of the time. Let me know if that sounds sensible and I'll update this PR to remove the `nav` and not test for `As` compatibility.

> we want to support sx on every component

No problem. I'll leave this in 🙂 

> we try to avoid breaking changes, but if there's a good case for them, we can schedule it for the next-major

No problem, I think there's no avoiding it here in any case (currently the HTML is invalid so we have to sort that out).

**Comment by siddharthkp** [comment]
Hi again! This looks good! 

I don't think we need to mark this as a breaking change. However, I did find 2 instances that we might want to check with canary build before we merge:

- https://github.com/github/memex/blob/main/src/client/components/view-navigation.tsx#L99
- https://github.com/github/docs-internal/blob/main/components/article/PlatformPicker.tsx#L163

**Comment by owenniblock** [comment]
Thanks for your attention to detail @siddharthkp!

For this one:

> in [memex/components/view-navigation.tsx#L99](https://github.com/github/memex/blob/main/src/client/components/view-navigation.tsx#L99), <TabNav role="tablist" would get attached to the base creating duplicate role=tablist

We're adding this change to fix an issue where the roles are incorrectly surfaced (because the `tablist` can only contain `tab` elements but currently the `tab` elements are wrapped in a `nav` so it created invalid semantic HTML.

> in [docs-internal/components/article/PlatformPicker.tsx#L163](https://github.com/github/docs-internal/blob/main/components/article/PlatformPicker.tsx#L163), data-test-id is passed down to the container and used for tests. With a quick look, this seems to be fine, but just to be sure, we should run the CI suite with a canary build

Sounds like a good idea! Is there a doc I can look at for how to do this? Or is it something you can help with?

**Comment by siddharthkp** [comment]
> We're adding this change to fix an issue where the roles are incorrectly surfaced (because the tablist can only contain tab elements but currently the tab elements are wrapped in a nav so it created invalid semantic HTML.

This would require a change in memex as well, right?

> Sounds like a good idea! Is there a doc I can look at for how to do this? Or is it something you can help with?

This isn't documented, I can write down the steps here as a start:

1. Grab the canary version from the pull request builds:
    image

2. Install canary version in a branch:

    ```
    npm install @primer/react@0.0.0-2022520124925
    ```

However, docs-internal is one major version behind (breaking changes), so I don't think we would be able to test this change in isolation :(

**Comment by siddharthkp** [comment]
> However, docs-internal is one major version behind (breaking changes), so I don't think we would be able to test this change in CI :(

Update: I made local changes in `node_modules` to test if tests will pass with the changes from this PR, they do :)

**Comment by owenniblock** [comment]
@siddharthkp I've created a draft PR here: https://github.com/github/memex/pull/10664 - I'll have to get some help to work out how best to test it!

**Comment by owenniblock** [comment]
OK, I'm getting an error on the `<TabNav.Link` line on memex - let me dig in to that and try and work out what's going on there!

**Comment by owenniblock** [comment]
@siddharthkp looks good with the canary build: https://github-1f17a67ba6-12535706.drafts.github.io/orgs/app/projects/1/views/1

Let me know when we should merge this (or if you think we need any further changes) 🙂

**Comment by siddharthkp** [comment]
Looks good, I'll merge it :)
